Sources/MessagingPush/PushClickHandler.swift

```diff
@@ -0,0 +1,220 @@
+    }
+
+    init(pushHistory: PushHistory, jsonAdapter: JsonAdapter) {
+        self.pushHistory = pushHistory
+        self.jsonAdapter = jsonAdapter
+    }
+
+    func setupClickHandling() {
+        // UNUserNotificationCenter.delegate is the 1 object in an iOS app that gets called when a push is clicked.
+        // We can set the CIO SDK as the delegate, but another SDK or the host app can set itself as the delegate instead which
+        // cause the CIO SDK to not be able to automatically handle pushes. To get around this we use swizzling to get notified when
+        // a new delegate gets set after we set the CIO SDK as the delegate. Therefore, we know when someone else takes over the
+        // click handling in the app.
+        swizzle(
+            targetClass: UNUserNotificationCenter.self,
+            targetSelector: #selector(setter: UNUserNotificationCenter.delegate),
+            myClass: PushClickHandlerImpl.self,
+            mySelector: #selector(PushClickHandlerImpl.cio_swizzled_setDelegate(delegate:))
+        )
+
+        if UNUserNotificationCenter.current().delegate == nil {
+            // Set our SDK as the click handler, if there isn't one already set in the app.
+            UNUserNotificationCenter.current().delegate = self
+        } else {
+            // This handles the case where a delegate may have already been assigned before our SDK is loaded into memory.
+            // This re-assignment triggers NotifiationCenter.delegate setter that we swizzled so our SDK can swizzle in its logic.
+            UNUserNotificationCenter.current().delegate = UNUserNotificationCenter.current().delegate
+        }
+    }
+
+    func setupClickHandling(onDelegate delegate: UNUserNotificationCenterDelegate) {
+        // Only swizzle on delegates that are not part of the CIO SDK.
+        // Problems such as infinite loops when a push is clicked can occur if the CIO SDK is setup as the app's click handler *and*
+        // we setup swizzling on our own delegate.
+        if delegate is PushClickHandlerImpl {
+            return
+        }
+
+        // Another SDK or the host app has set itself as the new UNUserNotificationCenter.delegate. We want to make sure the CIO SDK
+        // can still handle push clicks to make integration of the CIO SDK easy and reliable. We use swizzling on the new delegate instance
+        // so our SDK still gets notified when a push is clicked, even though the new delegate is setup to handle the push click.
+        swizzle(
+            targetClass: type(of: delegate),
+            targetSelector: #selector(UNUserNotificationCenterDelegate.userNotificationCenter(_:didReceive:withCompletionHandler:)),
+            myClass: PushClickHandlerImpl.self,
+            mySelector: #selector(PushClickHandlerImpl.cio_swizzle_didReceive(_:didReceive:withCompletionHandler:))
+        )
+    }
+
+    func pushClicked(response: UNNotificationResponse) -> CustomerIOParsedPushPayload? {
```

Comment:
This function does not yet compile as I am not yet done with this refactor. 

The idea here is to encapsulate the logic of click handling into 1 class to (1) prevent duplicate push metrics and deep link handling when a push is clicked (2) make this code more testable being in it's own class then `MessagingPush` extension functions, and (3) encapsulate automatic and manual push handling logic into 1 place in the code.